dstart prediction

WojciechMat opened the pull request:
I implemented dstart learning simply by quantizing dstart directly in DstartEncoder class.

WojciechMat committed:
dstart prediction

WojciechMat committed:
cosmetic changes

WojciechMat commented on train.py:
```diff
@@ -67,7 +80,7 @@ def train_model(
 ) -> nn.Module:
     # Get the index for padding token
     pad_idx = train_dataset.src_encoder.token_to_id["<blank>"]
-    src_vocab_size, tgt_vocab_size = vocab_sizes(cfg)
+    src_vocab_size, tgt_vocab_size = train_dataset.src_encoder.vocab_size, train_dataset.tgt_encoder.vocab_size
```
this makes the code more versatile

roszcz replied:
In my opinion having a separate way to calculate those values is better - you don't need to create a dataset to know those numbers, and the same line of code can be used in training, validation, dashboards and any other place where no "train_dataset" exist.

The design idea here is to have a function that acts as a "lens" to a problem - it's a term commonly used in functional programming, although it's much more complicated there :)

WojciechMat replied:
okay :)

WojciechMat commented on train.py:
```diff
@@ -15,12 +15,16 @@
 @hydra.main(version_base=None, config_path="configs", config_name="main")
 def main(cfg: DictConfig):
-    train_dataset, val_dataset = load_datasets(dataset_name=cfg.dataset_name, dataset_cfg=cfg.dataset)
+    train_dataset, val_dataset = load_datasets(
+        dataset_name=cfg.dataset_name,
+        dataset_cfg=cfg.dataset,
+        predict=cfg.predict,
+    )
     train_model(train_dataset, val_dataset, cfg)
```
I am struggling to think of a better way of passing the argument about the name of a column we want to predict and number of bins (always 200 right now) down to tgt_encoder in load_cache_dataset.

WojciechMat commented on data/tokenizer.py:
```diff
@@ -120,3 +126,54 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def tokenize(self, record: dict) -> list[str]:
+        # TODO I don't love the idea of adding tokens durint *tokenize* call
+        # If we want to pretend that our midi sequences have start and finish
+        # we should take care of that before we get here :alarm:
+
+        df = pd.DataFrame(record)
+
+        # Quantize only dstart
+        df["next_start"] = df.start.shift(-1)
+        df["dstart"] = df.next_start - df.start
+        df["dstart_bin"] = np.digitize(df.dstart.fillna(0), self.bin_edges) - 1
+
+        # get tokens from quantized data
+        tokens = [str(dstart_bin) for dstart_bin in df["dstart_bin"]]
+
+        tokens = ["<s>"] + tokens + ["</s>"]
+        return tokens
```
I decided to simply make quantization inside tokenize method, which does not seem very appropriate but does look clean and works well with our architecture.

roszcz replied:
I see the temptation of joining quantization with tokenization - maybe a single `Encoder` class could manage all of this in principle. But here, I'm also worried about the performance - I'm guessing that creating a new data frame for each sequence is going to slow down training a lot, and I don't want to cache data after tokenization.

Another problem here is that this function now does to separate steps, so if we wanted to keep this approach, it should be two functions `quantize` and `tokenize`.

WojciechMat replied:
Another approach I think, would be to apend "tgt_bins" key to our dataset. We would have to make the quantization back in our build_translation_dataset function and include it in quantized_piece_to_records sequence. 
We could then use it to handle dstart as well as duration, but it is redundant when using velocity (which is "quantized" into 128 bins :) ), but if we wanted to predict both dstart and duration at some point we might want to find a better way...
Or we can just create completly seperate program for dstart prediction but I do not think this is such a good idea.

WojciechMat reviewed (commented):
I left some comments to explain myself :P

roszcz committed:
add dashboard with velocity prompt experiments

WojciechMat committed:
bin edges

roszcz committed:
add prompt

roszcz commented:
@WojciechMat there are now 2 pull requests for the same task, this and #6 - can you clean this up?

roszcz committed:
show prompt after quantization

roszcz committed:
decrease repetition

roszcz committed:
add midi downloads

WojciechMat committed:
better dstart quantization

WojciechMat committed:
remove redundant copying

WojciechMat commented on data/tokenizer.py:
```diff
@@ -120,3 +126,52 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def tokenize(self, record: dict) -> list[str]:
+        # TODO I don't love the idea of adding tokens durint *tokenize* call
+        # If we want to pretend that our midi sequences have start and finish
+        # we should take care of that before we get here :alarm:
+        dstart = []
+        for it in range(len(record["start"]) - 1):
+            dstart.append(record["start"][it + 1] - record["start"][it])
+        dstart.append(0)
+
+        dstart_bins = np.digitize(dstart, self.bin_edges) - 1
+
+        # get tokens from quantized data
+        tokens = [str(dstart_bin) for dstart_bin in dstart_bins]
+        tokens = ["<s>"] + tokens + ["</s>"]
+        return tokens
```
I've managed to speed up the quantization during tokenization process 31 times on average by omitting converstion to pd.DataFrame and operating on dict instead.

WojciechMat commented on configs/main.yaml:
```diff
@@ -7,6 +7,8 @@ train:
   label_smoothing: 0.1
 
 dataset_name: 'roszcz/maestro-v1-sustain'
+predict: velocity
+tgt_bins: 200
```
not sure as to the "predict" name for this hyperparameter,
I also added tgt_bins to config to choose target values quantization bins number.

roszcz replied:
Maybe `target: velocity` would be better. `tgt_bins` also need a revamp - I think we should be explicit and call it `dstart_bins`. It could be set in the `dataset.quantization` section, and if you don't want to use the same MidiQuantizer approach, I think this should be a separate config

WojciechMat replied:
I could set dstart_bins in dataset.quantization if I was preparing a cached dataset with tgt_dstart_bins field. If I keep quantizng dstart during tokenization I would be caching the same object with different name (hashed config). We also do not need dstart_bins in datasets used for velocity prediction. BTW total quantization time in DstartEncoder is 2.75 seconds per epoch on train split of maestro-v1-sustain dataset so I think this is approach acceptable, what do you think?

roszcz replied:
How did you measure this encoder time?

Speed is one concern, so if it's fast then that's fine. My other concern is the code entanglement - I feel like we should rip the `dstart` and `velocity` predictions apart as much as possible 🤔 

Please take a look at https://github.com/Nospoko/autoencoders-demo and let me know how do you feel about doing a similar pipeline-based design in this project

WojciechMat replied:
I measured it by creating total_time atribute in DstartEncoder and used it to count time elapsed during quantization step inside tokenize function.

roszcz committed:
add a separate widget for piece selection, include file upload

roszcz committed:
cleanup

WojciechMat committed:
renamed hyperparameters

WojciechMat committed:
imports fix

WojciechMat committed:
imports fix

roszcz commented:
@WojciechMat Can you rebase this branch, so it doesn't show code already present on `master` as suggested changes? :)

WojciechMat committed:
clean rebase fail

WojciechMat committed:
clean rebase fail

WojciechMat commented:
Rebase done 🫡

WojciechMat commented:
We could do a pipeline-based design and use seperate configs for velocity, dstart and duration prediction. I would still quantize dstart in tokenize method though, as I find it to be the cleanest way. The training logic would be the same for velocity, dstart and duration, so the code for that could be shared.

roszcz commented:
> We could do a pipeline-based design and use seperate configs for velocity, dstart and duration prediction. I would still quantize dstart in tokenize method though, as I find it to be the cleanest way. The training logic would be the same for velocity, dstart and duration, so the code for that could be shared.

It's ok to do the quantization in the encoder, but right now your `tokenize` implementation does too much work doing first quantization, and then tokenization - I think in this form this will cause problems in the inference review, where we will need access to the quantized, but not tokenized values. Let me know if this makes sence!

WojciechMat commented:
> > We could do a pipeline-based design and use seperate configs for velocity, dstart and duration prediction. I would still quantize dstart in tokenize method though, as I find it to be the cleanest way. The training logic would be the same for velocity, dstart and duration, so the code for that could be shared.
> 
> It's ok to do the quantization in the encoder, but right now your `tokenize` implementation does too much work doing first quantization, and then tokenization - I think in this form this will cause problems in the inference review, where we will need access to the quantized, but not tokenized values. Let me know if this makes sence!

It does, but when reviewing model predictions we will need to untokenize them first as well. We could use the same function for src and tgt tokens to create piece with quantized dstart and predicted dstart and would not need to save additional tgt_dstart_bin to the dataset/samples.
If you want to have tgt_dstart_bin we could inject it to a dict pre-tokenization with some new DstartEncoder function, but I do not think we should change records inside the encoder.

roszcz commented:
What I think I'm getting at is that at some point we'll want to look at the input sequence and compare it to the target we're teaching the model to achieve. Looks to me that right now you'd have to do something like 

```py
# No
target_dstart = encoder.untokenize(encoder.tokenize(df.start))
```

but you should be able to do 

```py
# Yes
target_dstart = encoder.quantize(df.start)
```

WojciechMat committed:
quantize function

WojciechMat commented:
> What I think I'm getting at is that at some point we'll want to look at the input sequence and compare it to the target we're teaching the model to achieve. Looks to me that right now you'd have to do something like
> 
> ```python
> # No
> target_dstart = encoder.untokenize(encoder.tokenize(df.start))
> ```
> 
> but you should be able to do
> 
> ```python
> # Yes
> target_dstart = encoder.quantize(df.start)
> ```

done

WojciechMat committed:
pipeline architecture

WojciechMat commented on data/dataset.py:
```diff
@@ -116,6 +116,8 @@ def load_cache_dataset(
     dataset_cfg: DictConfig,
     dataset_name: str,
     split: str,
+    src_encoder: MidiEncoder,
+    tgt_encoder: MidiEncoder,
     force_build: bool = False,
 ) -> MyTokenizedMidiDataset:
```
what do you think of this way of preparing a dataset? (By passing src and tgt encoders to load_cache_dataset function)

roszcz replied:
I don't love it 💔 Maybe a better approach would be to move the `MyTokenizedMidiDataset` creation outside of this function, and only deal with `translation_dataset` here? 🤔

WojciechMat replied:
The dataset creation logic now looks like this:
```py
    src_encoder = QuantizedMidiEncoder(quantization_cfg=cfg.dataset.quantization)
    tgt_encoder = DstartEncoder(bins=cfg.dstart_bins)
    train_translation_dataset = load_cache_dataset(
        dataset_cfg=cfg.dataset,
        dataset_name=cfg.dataset_name,
        split="train",
    )
    val_translation_dataset = load_cache_dataset(
        dataset_cfg=cfg.dataset,
        dataset_name=cfg.dataset_name,
        split="validation",
    )

    train_dataset = MyTokenizedMidiDataset(
        dataset=train_translation_dataset,
        dataset_cfg=cfg.dataset,
        src_encoder=src_encoder,
        tgt_encoder=tgt_encoder
    )
    val_dataset = MyTokenizedMidiDataset(
        dataset=val_translation_dataset,
        dataset_cfg=cfg.dataset,
        src_encoder=src_encoder,
        tgt_encoder=tgt_encoder
    )

```
I think it's pretty good.

WojciechMat commented on evals/evaluate.py:
I have deleted depricated evaluate file and will implement pipeline-based evaluation logic.

WojciechMat committed:
update readme, clean

WojciechMat commented:
I have updated readme, added a dstart_conf, which you can choose by running
```
python train.py --config.name=dstart_conf
```
and changed the architecture to be pipeline-based.
Please let me know if that is what you expected.

WojciechMat committed:
MyTokenizedMidiDataset creation outside loading cached dataset

WojciechMat committed:
clean

WojciechMat committed:
fix configs

WojciechMat committed:
dstart prediction review

WojciechMat commented:
I've made sequence prediction review work for dstart predicting models and uploaded a checkpoint [here](https://huggingface.co/wmatejuk/midi-transformer-dstart)

WojciechMat committed:
folders for velocity and dstart checkpoints and dashboards

WojciechMat committed:
evaluation with greedy decode

WojciechMat committed:
evaluations and predict_piece dashboard

roszcz commented on configs/dstart_conf.yaml:
This file is already in a "configs" folder, so I'd say it's ok to just call it `dstart.yaml`. If you want to go the other direction, I would prefer not to be cheap on letters and call it `dstart_config.yaml` to make the reading experience smoother :)

roszcz commented on data/tokenizer.py:
```diff
@@ -120,3 +128,75 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+
+
+class DstartEncoder(MidiEncoder):
+    def __init__(self, bins: int = 200):
```
```suggestion
    def __init__(self, n_bins: int = 200):
```

`bins` suggests something else

roszcz commented on data/tokenizer.py:
```diff
@@ -120,3 +128,75 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def unquantized_start(self, dstart_bins: pd.Series) -> pd.Series:
```
```suggestion
    def unquantized_start(self, dstart_bins: np.array) -> np.array:
```

I think we don't need `pd.Series` here, numpy should make it more accessible

roszcz commented on data/tokenizer.py:
```diff
@@ -120,3 +128,75 @@ def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+        tokens = ["<s>"] + tokens + ["</s>"]
+        return tokens
+
+    def untokenize(self, tokens: list[str]) -> pd.DataFrame:
+        dstarts = [int(token) for token in tokens if token not in self.specials]
+        df = pd.DataFrame(dstarts, columns=["dstart_bin"])
+
+        return df
```
```suggestion
    def untokenize(self, tokens: list[str]) -> list[float]:
        dstarts = [int(token) for token in tokens if token not in self.specials]

        return dstarts
```

I'll revisit this after running the code more, but I'm guessing returning a list of floats here will make inference experiments easier to follow

WojciechMat replied:
How about we try and change it in VelocityEncoder as well then? Just so child objects implement methods the same way

roszcz replied:
Definitely!

WojciechMat replied:
QuantizedMidiEncoder untokenize would still need to return a dataframe though, as it untokenizes into four keys.

roszcz replied:
I think that's ok - it encodes and returns a complex structure so it seems justified

WojciechMat replied:
I found a problem - righ now we take greedy decode output - which is baisically output token ids - and use untokenize function as if they were tokens (strings)
We should convert output to tokens first or create a function to decode token ids directly.

roszcz replied:
Sorry I don't follow - how can we use untokenize on token ids as if they were tokens?

WojciechMat replied:
Right now we do smtm like this:
Model output --untokenize—> sequence (all in generate output function)

We should do is
Model output -> vocab[id] for id in model output -> untokenize -> sequence

I do not know why when passing tensor of ints into untokenize we do not get error with type hints, as the argument for untokenize should be a list of strings

roszcz replied:
Can you link to a place in the code where this is happening? I just took a random look at one of the velocity dashboards `dashboard/velocity/predict_piece.py:77`, and there is vocab after model:

```py
out_tokens = [tgt_encoder.vocab[x] for x in predicted_token_ids if x != pad_idx]
```

WojciechMat replied:
yup, it works great here

WojciechMat committed:
changed datatypes and parameter name in DstartEncoder

WojciechMat committed:
VelocityEncoder changed datatype, fixed generate sequence

WojciechMat commented on dashboard/dstart/predict_piece.py:
```diff
@@ -0,0 +1,141 @@
+            start_symbol=0,
+            device=train_cfg.device,
+        )
+
+        out_tokens = [tgt_encoder.vocab[x] for x in predicted_token_ids if x != pad_idx]
```
here it works okay

WojciechMat commented on dashboard/dstart/predict_piece.py:
```diff
@@ -0,0 +1,141 @@
+            max_len=128,
+            start_symbol=0,
+            device=train_cfg.device,
+        )
+
+        out_tokens = [tgt_encoder.vocab[x] for x in predicted_token_ids if x != pad_idx]
+        predicted_tokens += out_tokens
+
```
here it works okay

WojciechMat commented on dashboard/dstart/main.py:
```diff
@@ -0,0 +1,218 @@
+        record_source = json.loads(record["source"])
+        src_token_ids = record["source_token_ids"]
+
+        generated_dstart = generate_sequence(
+            model=model,
+            device=DEVICE,
+            pad_idx=pad_idx,
+            src_tokens=src_token_ids,
+            tgt_encoder=dataset.tgt_encoder,
+            sequence_size=train_cfg.dataset.sequence_len,
+        )
```
here it does not

WojciechMat commented on utils.py:
```diff
@@ -95,6 +101,7 @@ def generate_sequence(
         device=device,
     )
 
+    sequence = [tgt_encoder.vocab[idx] for idx in sequence]
     out_sequence = tgt_encoder.untokenize(sequence)
```
It did not work inside generate_output function (used in dashboard/velocity/main.py and dashboard/dstart/main.py, because we were missing this line of code.

WojciechMat replied:
It works okay now, but we have old version on master rn

roszcz replied:
Please don't treat github comments like code comments - I'm not able to follow what the issue you're pointing to is with multiple comments across multiple files. If there's an issue, just comment on the line that is causing the issue and reference other sections of the code if necessary :)

Actually, you shouldn't do that with code comment either 😅

WojciechMat replied:
Sorry :)

WojciechMat commented:
I implemented you requests and fixed generate_sequence, if you do not have any more suggestions I think the pull is ready :rocket:

WojciechMat committed:
Update README.md

WojciechMat committed:
changed variable name

WojciechMat committed:
Merge branch 'MIDI-74/new-dstart' of github.com:Nospoko/midi-translation into MIDI-74/new-dstart
